apps/studio/src/lib/db/clients/bigquery/BigQueryCursor.ts

```diff
@@ -0,0 +1,111 @@
+import { BeeCursor } from "../../models";
+import rawLog from 'electron-log';
+import { BigQuery, GetRowsOptions, Job, QueryResultsOptions, RowMetadata } from "@google-cloud/bigquery";
+import { waitFor } from "../base/wait";
+
+const log = rawLog.scope('bigquerycursor');
+type Conn = BigQuery;
+
+interface MiniCursor {
+  options: QueryResultsOptions
+  job: Job
+}
+
+export class BigQueryCursor extends BeeCursor {
+
+  private cursor?: MiniCursor;
+  private rowBuffer: any[][] = [];
+  private bufferReady = false;
+  private end = false;
+  private error?: Error;
+
+  constructor(
+    private conn: Conn, 
+    private query: string, 
+    private params: string[], 
+    chunkSize: number
+  ) {
+    super(chunkSize);
+  }
+
+  async start(): Promise<void> {
+    log.debug("start");
+
+    const query = {
+      query: this.query,
+      params: this.params
+    }
+    const [job] = await this.conn.createQueryJob(query);
+    const promise = new Promise<void>((resolve, _reject) => {
+      const queryOptions = {
+        autoPaginate: false,
+        // wtf bigquery?
+        startIndex: '0',
+        maxResults: this.chunkSize
+      }
+
+      job.getQueryResults(queryOptions, this.handleRow.bind(this));
+      // job.on('complete', this.handleEnd.bind(this));
+      job.on('error', this.handleError.bind(this));
+      this.cursor = { job: job, options: queryOptions };
+      resolve();
+    })
+    return promise;
+  }
+
+  private handleRow(err: Error | null, resource?: RowMetadata, nextQuery?: GetRowsOptions, _response?: any): void {
+    if (err) {
+      this.handleError(err);
+      return;
+    }
+    if (nextQuery) {
+      this.cursor.options.pageToken = nextQuery.pageToken;
+    } else {
+      this.handleEnd();
+    }
+    this.rowBuffer.push(...resource);
+    this.pause();
+  }
+
+  private handleError(error: Error) {
+    log.debug("handling error");
+    this.error = error;
+    console.error(error);
+  }
+
+  private handleEnd() {
```

Comment:
this never fires right now

Reply:
It does fire, it's just not fired from a listener. It's fired from `handleRow`